Made `remove_dir_all` work as documented when directory is a symlink

tailhook opened the pull request:
Previously the function worked correctly only when symlink is inside of
a directory, not when path is a symlink itself.

This patch also eliminates the `lstat()` call on most systems because
`DirEntry::file_type` is enough both to detect if entry is a dir and if
entry is a symlink.

I'm not sure if the ``#[cfg(not(windows))]`` and the comment above is still actual. Can't test patch on windows.

This supercedes #29324

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @brson (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way Github handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

eefriedman commented on src/libstd/fs.rs:
```diff
@@ -1069,13 +1069,23 @@ pub fn remove_dir_all<P: AsRef<Path>>(path: P) -> io::Result<()> {
 fn _remove_dir_all(path: &Path) -> io::Result<()> {
+    let filetype = try!(symlink_metadata(path)).file_type();
+    if filetype.is_symlink() {
+        remove_file(path)
```
Not sure this will work correctly on Windows; a symlink isn't always a file on Windows.

tailhook replied:
Well, the code that removed a symlink inside the directory used `remove_file` too, AFAICS. So at least it's not a regression.

Anyway, it would be helpful if someone can test it on windows.

alexcrichton replied:
I've confirmed that directory symlinks on windows cannot be removed by `remove_file` but can be removed via `remove_dir`.

alexcrichton replied:
I think that this may want to also use `is_dir` to choose whether to recurse or not (like below). Junctions on Windows (e.g. hard links to directories) don't report `is_dir`, but to behave the same as unix we should delete their contents.

tailhook replied:
> I think that this may want to also use is_dir to choose whether to recurse or not (like below). Junctions on Windows (e.g. hard links to directories) don't report is_dir, but to behave the same as unix we should delete their contents.

Isn't it a contradiction? If they don't report `is_dir` how should we know to recurse?

The question probably is: `is_dir() && is_symlink()` (impossible on unix) should recurse or `remove_dir()` ?

tailhook replied:
Okay, it looks like impossible combination on windows too:
https://github.com/rust-lang/rust/blob/master/src/libstd/sys/windows/fs.rs#L429

So probably I must stat symlink on windows to find out whether it's a directory. And then if it's a directory do (non-recursive) `remove_dir`. Right? According to wikipedia, junctions are removed with `rmdir` (i.e. non-recursive)

eefriedman replied:
`remove_dir()` seems like the answer which is most consistent with other forms of symbolic links.

alexcrichton replied:
Hm yes I think I may have gotten a little confused, I'll have to think on this a bit.

eefriedman commented on src/libstd/fs.rs:
```diff
@@ -1724,6 +1734,23 @@ mod tests {
     }
 
+    // FIXME(#12795) depends on lstat to work on windows
+    #[cfg(not(windows))]
```
Is it really necessary to disable this test on Windows?  #12795 was fixed a long time ago.

tailhook replied:
As I outlined in the description of the pull request, I don't know. Just copied it from another test on remove_dir_all. Can you test on windows?

eefriedman replied:
Oh, nevermind, the test won't work because soft_link doesn't do the right thing on Windows.

Actually, given that this is a new test, please avoid using `soft_link` because it's deprecated.

tailhook replied:
So should I remove the comment but leave cfg(not(windows))? (so use symlink from os::unix)

I'm pretty sure tests for windows should be separate testing both symlink_file, and symlink_dir. And I'm unable to write them.

eefriedman replied:
Yes, that's fine for now; once this is merged, you can file a followup issue about adding Windows tests.

nagisa commented on src/libstd/fs.rs:
```diff
@@ -1069,13 +1069,23 @@ pub fn remove_dir_all<P: AsRef<Path>>(path: P) -> io::Result<()> {
+        _remove_dir_all_unchecked(path)
+    }
+}
+fn _remove_dir_all_unchecked(path: &Path) -> io::Result<()> {
```
Any real reason to have this function split off? It is not big enough to justify splitting it for stylistic reasons IMO.

alexcrichton replied:
It's fine to remove the leading underscore here, the convention above is just because it's paired with a sibling `remove_dir_all` function.

alexcrichton replied:
Also this may be best with a `_recursive` prefix here instead of `_unchecked`

nagisa commented:
I’d like to see a test which tests whether function behaviour is as expected. A “make-run” test probably would be the least painful one to do.

tailhook commented:
> I’d like to see a test which tests whether function behaviour is as expected.

There is a test, function called "recursive_rmdir_of_symlink"

> A “make-run” test probably would be the least painful one to do.

Well, what is a "make-run" test?

nagisa commented:
> There is a test, function called "recursive_rmdir_of_symlink"

Ah, ignore me then. I should go to sleep instead of reviewing more PRs, I guess.

alexcrichton commented on src/libstd/fs.rs:
```diff
@@ -1069,13 +1069,23 @@ pub fn remove_dir_all<P: AsRef<Path>>(path: P) -> io::Result<()> {
-            try!(remove_dir_all(&*child));
+        let child = try!(child);
+        let child_path = child.path();
+        let mut child_type = try!(child.file_type());
```
This should fall back to calling `symlink_metadata` in the case `file_type()` failed (as it's not guaranteed to work across all platforms right now)

tailhook replied:
AFAICS, fallback is already inside DirEntry: https://github.com/rust-lang/rust/blob/master/src/libstd/sys/unix/fs.rs#L186 (unix), on windows it looks like it always work.

alexcrichton replied:
Oh, right! I forgot about that, carry on!

tailhook commented:
Okay, I've renamed the method and added some windows support. Unfortunately, windows support is untested. But I've included few provisional tests for windows. Hopefully @bors will run the tests for me.

Speaking about windows: I do `metadata()` for symlinks, to find out which function to use for delete. Unit tests are for symlinked files and directories. Junctions should use `remove_dir`, so should work too in theory.

retep998 commented:
Just so you know, you cannot actually create any symbolic links in the tests because that requires administrator privileges on Windows for some obscure reason. Directory junctions and hard links are fine, just not symbolic links.

tailhook commented:
Well, so I should remove the tests back again. @retep998, does `std::fs::hard_link` create a junction when destination is a directory? I may probably just replace symlinks into hard_links/junctions in tests.

retep998 commented:
Hard links are for files only while junctions are for directories only. Also hard links are not reparse points.

tailhook commented:
@retep998, any way to create a junction in rust?

retep998 commented:
https://github.com/rust-lang/rust/blob/master/src/libstd/sys/windows/fs.rs#L652-L690

alexcrichton commented on src/libstd/fs.rs:
```diff
@@ -1069,13 +1069,43 @@ pub fn remove_dir_all<P: AsRef<Path>>(path: P) -> io::Result<()> {
-            try!(remove_file(&*child));
+            if cfg!(windows) {
+                if child_type.is_symlink() {
+                    let target_type = try!(metadata(&*child_path)).file_type();
```
I tested this out a bit, and oddly enough I'm not sure we can use this function to determine this. Using `metadata` will resolve all intermediate symlinks, showing the file type of the actual destination file. This ends up having what I think are two problems:

1. If the symlink is broken (e.g. the destination is gone), I don't think this will work.
2. Removal of the symlink depends on what kind of symlink was created, not the destination itself.

I found that I could create a symlink to a file with `symlink_dir`, but I could only remove the symlink with `remove_dir`. I think that we may have enough information in the `DirEntry` itself to figure this out (via a private API for now).

retep998 replied:
>I found that I could create a symlink to a file with `symlink_dir`,

That's because Windows doesn't check the target of symbolic links when you create them. You still created a directory symbolic link. It just happens to be pointing at an invalid thing.

alexcrichton commented on src/libstd/fs.rs:
```diff
@@ -1717,13 +1747,83 @@ mod tests {
+        assert!(canary.exists());
+    }
+
+    #[cfg(windows)]
```
I think it's fine to include these tests, they'll need to be guarded though to see if the current user has the privilege to create symbolic links. I don't think we want to require the test suite to run as an administrator by default, unfortunately, but I don't mind going into the bots and giving the rustbuild user privileges to make symlinks!

alexcrichton replied:
And also while you're at it, could you add test cases like:

* `remove_dir_all` with a broken symlink inside
* remove a directory with a directory symlink to a file 
* remove a directory with a file symlink to a directory

tailhook replied:
> they'll need to be guarded though to see if the current user has the privilege to create symbolic links

Any hints on how to do that?

alexcrichton replied:
I think the [`PrivilegeCheck`](https://msdn.microsoft.com/en-us/library/windows/desktop/aa379304%28v=vs.85%29.aspx) API looks like it may do the trick? (I haven't tested this out though).

alexcrichton commented on src/libstd/fs.rs:
```diff
@@ -1069,13 +1069,43 @@ pub fn remove_dir_all<P: AsRef<Path>>(path: P) -> io::Result<()> {
+        } else {
+            // On unix symlinks are always files
+            remove_file(path)
+        }
```
This might be good to extract into a helper function to be shared with below, something like:

```rust
fn remove(path: &Path, meta: &fs::Metadata) -> io::Result<()> {
    // ...
}
```

steveklabnik commented:
What's the status of this PR?

tailhook commented:
Well, it works on linux. But it seems I don't have a time to work on windows stuff. And as I've already written I can't test anyway.

bors commented:
:umbrella: The latest upstream changes (presumably #31360) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
cc @pitdicker 

you may be interested in this as well!

pitdicker commented:
Yes I am :)

pitdicker commented:
@tailhook I feel bad but I hijacked your pr.

tailhook commented:
@pitdicker, never mind. I'm happy someone fixes it :) Closing in favor of #31468
